Changed autoCapitalize attribute from string/boolean to specific values

ryo-manba opened the pull request:
This PR updates the autoCapitalize attribute type in React from string | boolean to specific string values: "off", "none", "on", "sentences", "words", and "characters", following the HTML standard.

Reference: [HTML autocapitalize attribute](https://html.spec.whatwg.org/multipage/interaction.html#attr-autocapitalize)

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:

- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:

- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

If removing a declaration:

- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

eps1lon commented on types/react-native-web/index.d.ts:
```diff
@@ -1093,7 +1093,7 @@ export interface TextProps extends ViewPropsWithoutStyle {
 
 export interface TextInputProps extends ViewPropsWithoutStyle {
-    autoCapitalize?: "characters" | "none" | "sentences" | "words";
+    autoCapitalize?: "off" | "none" | "on" | "sentences" | "words" | "characters" | undefined;
```
Are we sure `react-native-web` supports all of these values?

eps1lon commented on types/react-native-material-ui/index.d.ts:
```diff
@@ -427,7 +427,7 @@ export interface ToolbarStyle {
     placeholder?: string | undefined;
     autoFocus?: boolean | undefined;
-    autoCapitalize?: "none" | "sentences" | "words" | "characters" | undefined;
+    autoCapitalize?: "off" | "none" | "on" | "sentences" | "words" | "characters" | undefined;
```
Are we sure `react-native-material-ui` supports all these values? If it just forwards this prop, it should just reference the original type instead of inlining it.

eps1lon commented on types/react-mdl/index.d.ts:
```diff
@@ -33,6 +33,7 @@ declare namespace __ReactMDL {
         allowTransparency?: boolean | undefined;
         alt?: string | undefined;
         async?: boolean | undefined;
+        autoCapitalize?: "off" | "none" | "on" | "sentences" | "words" | "characters" | undefined;
```
Same here. Why not reference the original type?

eps1lon commented on types/notie/index.d.ts:
```diff
@@ -66,7 +66,7 @@ export interface InputOptions {
     position?: Position | undefined;
     /** @default 'none' */
-    autocapitalize?: string | undefined;
+    autoCapitalize?: "off" | "none" | "on" | "sentences" | "words" | "characters" | undefined;
```
Why not reference the original type?

eps1lon reviewed (changes requested):
Thank you. Did we restrict enumerated props in the past or did we include a `| (string & {})` fallback to support existing usages?

We also need to double check if this change should land in all libraries.

ryo-manba commented:
@eps1lon 
Thank you for your comment. I have removed the unnecessary package changes.
It appears that this has been included since React 16, so I have updated it accordingly.
ref: https://github.com/facebook/react/pull/10564

In that PR, they added a `|(string & {})` fallback. Should I do the same here?
ref:  https://github.com/DefinitelyTyped/DefinitelyTyped/pull/53361

eps1lon commented:
> In that PR, they added a |(string & {}) fallback. Should I do the same here?

Let's go with that for now.

eps1lon commented:
Thank you
